Fix Don't select first item when multiple is true

ItWorksOnMyMachine opened the pull request:
When multipe is true on the select, the first item shouldn't be selected by default.

ItWorksOnMyMachine committed:
Fix Don't select first item when multiple is true

When multipe is true on the select, the first item shouldn't be selected by default.

ghiscoding commented on aurelia-bootstrap-select/src/abp-select.js:
```diff
@@ -394,10 +394,10 @@ export class AbpSelectCustomElement {
       this.selectedValue = (this.util.isObject(this.collection[0]) ? this.collection[0][this.objectKey] : this.collection[0]);
       this.selectedItem = this.collection[0];
-    } else {
+    } else if (!this.isEmptySelection(selection)) {
```
not sure about this line change, if there's nothing selected then `selectedValue` and `selectedItem` might contain the previous values. I'm just reading the code, so I might be wrong in my assumptions

ItWorksOnMyMachine replied:
Yeah, good call.  I'll make an update.

ItWorksOnMyMachine committed:
Fix Update to support multiple and null initial values

My update to renderSelection was just plain wrong.  And renderPreSelection now accounts for empty values where multiple is true and false separately.  I think this should work.

ghiscoding commented:
@ItWorksOnMyMachine 
Can you pull latest code that I just merged in `master` and merge it in your branch. I'd like to make sure your code is up to date. Thanks

@Mobe91 
You have a lot more experiences with this lib that I do, would you mind taking a quick look at this PR? I would feel much safer after your review. Thanks a lot

Mobe91 commented:
@ItWorksOnMyMachine So what you want to achieve is that even if `emptyOnNull` is `false`, no select items should be pre-selected in case of a multi-select? May I ask why?

Apart from that I noticed that `emptyOnNull` used to work as intended for multi-select elements before 1.4.0, i.e. no collection item was pre-selected if no value was set and `emptyOnNull` was set to `true`. However, from my current understanding this was only working by accident and stopped working with https://github.com/ghiscoding/Aurelia-Bootstrap-Plugins/commit/f298e002fd6cb77dbce1ea96e4f72ee9c5b00ab2#diff-98305bbb74566ae710d271bb4aa03893R456. @ghiscoding Do you know the reason for this change? Further, there are some inconsistencies in the implementation that need to be fixed in order to get that right I think. This starts with aligning the rendered selection with the `selectedValue` and `selectedItem` in `renderPreSelection` (not tested):
```javascript
this.selectedValue = selection.index = (this.util.isObject(this.collection[0]) ? this.collection[0][this.objectKey] : this.collection[0]);
this.selectedItem = selection.item = this.collection[0];
```

ghiscoding commented:
@Mobe91 
The question you asked about the commit for version `1.4.0`, was all related to get the Optgroup working with 1 or more items, apparently it wasn't working at all when only 1 item was in the Optgroup.

I have a hard time to test all of these changes and my usage is usually rather simple. It would be nice to have unit test, but sadly I have no personal time to dedicate to that.

ItWorksOnMyMachine commented:
I think I misunderstood how emptyOnNull was supposed to be working, mainly because it wasn't working as I thought it should be.  My change was intended to make the render functions more consistent with how selectedItemChanged uses multiple and emptyOnNull.  I think I failed at that.  Ultimately, my issue is that when renderPreSelection is called, if selection is empty, then selectedValue/selectedItem is set to the first element in the collection.  This is wrong for multi-select lists.  There is a valid scenario where nothing is selected, and the functionality as it stands is making it impossible for my customers to empty a filter on a page and get all results returned from the server.  The only workaround is to open the dialog and select everything.  But then, selecting everything by clicking the "Select all" button, was causing the infinite loop because bootstrap-select was jumbling the order on each call of refresh.

ghiscoding commented:
I might be missing documentation on the `emptyOnNull` but anyway my implementation at first was: "On the 1st pass **only** (page load), if the value is `null` then don't pre-select anything in the select). We could maybe update the README with an explanation. I'm also not sure if latest code has this behavior with multi-select though, I might have forgot to test that part. 

Anyhow, this can be seen in the demo (note that the demo might not have latest version)
![emptyonnull](https://user-images.githubusercontent.com/643976/47108254-36310e00-d219-11e8-9ffb-6ed2ce70628e.png)

Mobe91 commented:
@ItWorksOnMyMachine 
> Ultimately, my issue is that when renderPreSelection is called, if selection is empty, then selectedValue/selectedItem is set to the first element in the collection.

This is exactly what setting `emptyOnNull` to `true` is supposed to prevent. But this feature is broken in current master and most likely also in the latest 1.4.0 release.

I supplied PR #77 that should fix this. With this change, the demo by @ghiscoding works as shown in the screenshot - without this fix, the highlighted select box would have the first collection element selected although `emptyOnNull` is `true`.

@ghiscoding I suggest to close this PR in favor of #77 .

ghiscoding commented:
@Mobe91 
Thanks a lot for the review and the PR to fix this, I will review and test it in the coming days, we should expect a new version by the end of the weekend. I will probably use a patch version `1.4.1` unless you think it requires a minor version.

ItWorksOnMyMachine commented:
@mobe91, thanks for the help! @ghiscoding, patch seems right to me.

ghiscoding commented:
Closing in favor of PR #77
Thanks for all feedback and help from @Mobe91 

Released under version `1.4.1`, also updated demo
Thanks again
